proposals/dictionary-expressions.md

```diff
@@ -231,22 +293,36 @@ This would allow `ImmutableDictionary<TKey, TValue>` to be annotated with `[Coll
 
 We could consider not adding this rule, and instead still require the create method to take a `ReadOnlySpan<E>`.  However, that would require the BCL to then add such a method to `ImmutableDictionary` as the existing `CreateRange` methods take an `IEnumerable`.
 
-### Open question 2
+Conclusion: TBD.  Working group recommendation.  We will defer to the BCL here.  They will have to add `[CollectionBuilder]` annotations to light up anyways.  They can either point these at the existing `CreateRange` methods, or they could point at new methods that take spans.
 
-It is common for dictionaries to take in a `comparer` value, to determine how keys are compared when adding, removing, and looking them up.  Use in the ecosystem is prevalent, and much discussion and feedback from the community is that being able to supply a comparer is important to them.  How could we accomplish this with the new dictionary-expression?  Options include:
+Pointing at the existing method would have the positive that it would add to the BCL surface area.  But it would have the negative of forcing going through interfaces to create an immutable dictionary.
 
-1. Provide no support (or no support in C# 13), leaving such dictionaries as ones you could not use *dictionary expressions* for.
+Pointing at a new method would allow for more efficient, stack-based, initialization.  However, it would also increase the BCL surface area.
 
-1. Provide a special syntactic form *solely* for the purpose of supplying that value.  For example: `[comparer: myComp, "mads": 21, .. ldmMembers]`.  Here `comparer` would be a contextual keyword.  Users wanting to use that as an actual key would say `@comparer: 1`
+BCL might also prefer something like `CreateRange(ReadOnlySpan<TKey> keys, ReadOnlySpan<TValues> values)` such a signature would then allow constant data segments to be pointed at with an dictionary expression like so:
 
-1. Provide a special syntactic form for the purpose of supplying data to constructors and create methods.  For example: `[new: (comparer: myComp, capacity: 50), "mads": 21, .. ldmMembers]`
+```c#
+FrozenDictionary<int, int> mapping = [1: 2, 2: 4, 4: 8];
+
+// maps to:
+// No stack allocs at all!
+private static ReadOnlySpan<int> __keys => [1, 2, 4];
+private static ReadOnlySpan<int> __values => [2, 4, 8];
+mapping = FrozenDictionary.CreateRange(__keys, __values);
+```
```

Comment:
@stephentoub for thoughts. Specifically, would you prefer we don't change the surface area of the BCL and the compiler just redirects to the existing CreateRange methods that take an IEnumerable<KVP>? Or is there appetite for more efficient overloads that you would add that woudl allow efficient stack-allocing of keys and values (or KVPs), and passing those along?

We defer to you here.

Reply:
Adding new API surface area is fine, albeit preferably API surface area that makes sense even without the language feature.

In this particular case, what's the benefit of a create method that takes a `ROS<TKey>` + `ROS<TValue>` rather than one that takes an `ROS<KeyValuePair<TKey,TValue>>`?

Reply:
I think we aren't sure and are looking for input, but the potential thought was that the spans of key or value, when primitive and constant, could come from the data segment, or something like that. KVP might rule out such an approach for all we know.